src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.SizeOpt.cs

```diff
@@ -0,0 +1,109 @@
+
+
+            ReadOnlySpan<char> input = new ReadOnlySpan<char>(pInputBuffer, inputLength);
+            Span<byte> output = new Span<byte>(pOutputBuffer, outputBytesRemaining);
+
+            OperationStatus opStatus = OperationStatus.Done;
+            while (!input.IsEmpty)
+            {
+                opStatus = Rune.DecodeFromUtf16(input, out Rune rune, out int charsConsumedJustNow);
+                if (opStatus != OperationStatus.Done) { break; }
+                if (!rune.TryEncodeToUtf8(output, out int bytesWrittenJustNow)) { opStatus = OperationStatus.DestinationTooSmall; break; }
+                input = input.Slice(charsConsumedJustNow);
+                output = output.Slice(bytesWrittenJustNow);
+            }
+
+            pInputBufferRemaining = pInputBuffer + inputLength - input.Length;
+            pOutputBufferRemaining = pOutputBuffer + outputBytesRemaining - output.Length;
+
+            return opStatus;
+        }
+
+        // Returns &inputBuffer[inputLength] if the input buffer is valid.
+        /// <summary>
+        /// Given an input buffer <paramref name="pInputBuffer"/> of byte length <paramref name="inputLength"/>,
+        /// returns a pointer to where the first invalid data appears in <paramref name="pInputBuffer"/>.
+        /// </summary>
+        /// <remarks>
+        /// Returns a pointer to the end of <paramref name="pInputBuffer"/> if the buffer is well-formed.
+        /// </remarks>
+        /// <param name="pInputBuffer">Pointer to Utf8 byte buffer</param>
+        /// <param name="inputLength">Buffer length in bytes</param>
+        /// <param name="utf16CodeUnitCountAdjustment">Zero or negative number to be added to the "bytes processed" return value to come up with the total UTF-16 code unit count.</param>
+        /// <param name="scalarCountAdjustment">Zero or negative number to be added to the "total UTF-16 code unit count" value to come up with the total scalar count.</param>
+        public static byte* GetPointerToFirstInvalidByte(byte* pInputBuffer, int inputLength, out int utf16CodeUnitCountAdjustment, out int scalarCountAdjustment)
+        {
+            Debug.Assert(inputLength >= 0, "Input length must not be negative.");
+            Debug.Assert(pInputBuffer != null || inputLength == 0, "Input length must be zero if input buffer pointer is null.");
+
+            var input = new ReadOnlySpan<byte>(pInputBuffer, inputLength);
+            int cumulativeUtf16CodeUnitCount = 0;
+            int cumulativeBytesConsumed = 0;
+            int cumulativeScalarValueCount = 0;
+            while (!input.IsEmpty)
+            {
+                if (Rune.DecodeFromUtf8(input, out Rune rune, out int bytesConsumed) != OperationStatus.Done)
+                    break;
+                input = input.Slice(bytesConsumed);
+                cumulativeUtf16CodeUnitCount += rune.Utf16SequenceLength;
+                cumulativeBytesConsumed += bytesConsumed;
```

Comment:
You don't need this counter. You can calculate bytes consumed after the loop as "original input length - remaining length of 'input' span".